src/map/clif.c

```diff
@@ -8099,65 +8099,47 @@ static void clif_devotion(struct block_list *src, struct map_session_data *tsd)
 		clif->send(buf, packet_len(0x1cf), src, AREA);
 }
 
- /**
-  * Server tells clients nearby 'sd' (and itself) to display spirits spheres
-  * Notifies clients in an area or self of an object's spirits.
-  * 01d0 <id>.L <amount>.W (ZC_SPIRITS)
-  * 01e1 <id>.L <amount>.W (ZC_SPIRITS2)
-  *
-  * @param bl     Source block list.
-  * @param spirit Type of spirit data from sd.
-  * @param target Either target is AREA or SELF.
-  */
-static void clif_spiritball(struct block_list *bl, enum spirit_ball_types spirit, enum send_target target)
+/**
+ * Server tells clients nearby 'sd' (and itself) to display soulballs.
+ * Notifies clients in an area or self of an object's soulballs.
+ * 01d0 <id>.L <amount>.W (ZC_SPIRITS)
+ *
+ * @param bl        Source block list.
+ * @param soulballs amount of soulballs
+ * @param target    Either target is AREA or SELF.
+ */
+static void clif_soulball(struct block_list *bl, int soulballs, enum send_target target)
 {
 	nullpo_retv(bl);
 
 	struct PACKET_ZC_SPIRITS p = { 0 };
 
 	p.PacketType = HEADER_ZC_SPIRITS;
 	p.AID = bl->id;
-	p.num = 0;
-	switch (bl->type) {
-	case BL_PC:
-	{
-		struct map_session_data *sd = BL_CAST(BL_PC, bl);
-		nullpo_retv(sd);
+	p.num = soulballs;
+	clif->send(&p, sizeof(struct PACKET_ZC_SPIRITS), bl, target);
+}
 
-		switch (spirit) {
-		case BALL_TYPE_SPIRIT:
-			p.num = sd->spiritball;
-			break;
-		case BALL_TYPE_SOUL:
-			p.AID = sd->bl.id;
-			p.num = sd->soulball;
-			break;
-		case BALL_TYPE_NONE:
-			break;
-		}
-		clif->send(&p, sizeof(struct PACKET_ZC_SPIRITS), ((bl == NULL && spirit == BALL_TYPE_SOUL) ? &sd->bl : bl), (spirit == BALL_TYPE_SPIRIT ? AREA : target));
```

Comment:
the logic here in previous version is somewhat confusing to me, but:
1. `nullpo_retv(bl);` at the start, so `(bl == NULL && ..)` is always false, thus `bl` is always the target (Well, it would be anyway)
2. the `target` for SPIRIT is always `AREA`, this is given by the caller already.


so simply changing to `bl, target` is equivalent.

Reply:
1. I agree
2. `target `is not always `AREA`